lib/ProductOpener/Images.pm

```diff
@@ -1996,93 +1996,146 @@ sub extract_text_from_image ($product_ref, $id, $field, $ocr_engine, $results_re
 		}
-
 	}
 	elsif ($ocr_engine eq 'google_cloud_vision') {
 
-		my $url = "https://vision.googleapis.com/v1/images:annotate?key="
-			. $ProductOpener::Config::google_cloud_vision_api_key;
+		my $json_file = "$www_root/images/products/$path/$filename.json";
+		open(my $gv_logs, ">>:encoding(UTF-8)", "$data_root/logs/cloud_vision.log");
+		my $cloudvision_ref = send_image_to_cloud_vision($image, $json_file, $gv_logs);
+		close $gv_logs;
 
-		my $ua = LWP::UserAgent->new();
+		if (    (defined $cloudvision_ref->{responses})
+			and (defined $cloudvision_ref->{responses}[0])
+			and (defined $cloudvision_ref->{responses}[0]{fullTextAnnotation})
+			and (defined $cloudvision_ref->{responses}[0]{fullTextAnnotation}{text}))
+		{
 
-		open(my $IMAGE, "<", $image) || die "Could not read $image: $!\n";
-		binmode($IMAGE);
-		local $/;
-		my $image_data = do {local $/; <$IMAGE>};    # https://www.perlmonks.org/?node_id=287647
-		close $IMAGE;
+			$log->debug("text found in google cloud vision response") if $log->is_debug();
 
-		my $api_request_ref = {
-			requests => [
-				{
-					features => [{type => 'TEXT_DETECTION'}],
-					# image => { source => { imageUri => $image_url}}
-					image => {content => encode_base64($image_data)}
-				}
-			]
-		};
-		my $json = encode_json($api_request_ref);
+			$results_ref->{$field} = $cloudvision_ref->{responses}[0]{fullTextAnnotation}{text};
+			$results_ref->{$field . "_annotations"} = $cloudvision_ref;
+			$results_ref->{status} = 0;
+			$product_ref->{images}{$id}{ocr} = 1;
+			$product_ref->{images}{$id}{orientation}
+				= compute_orientation_from_cloud_vision_annotations($cloudvision_ref);
+		}
+		else {
+			$product_ref->{images}{$id}{ocr} = 0;
+		}
+	}
+	return;
+}
 
-		my $request = HTTP::Request->new(POST => $url);
-		$request->header('Content-Type' => 'application/json');
-		$request->content($json);
+sub send_image_to_cloud_vision ($image_path, $json_file, $gv_logs) {
 
-		my $res = $ua->request($request);
-		# $log->info("google cloud vision response", { json_response => $res->decoded_content, api_token => $ProductOpener::Config::google_cloud_vision_api_key });
+	my $url
+		= "https://vision.googleapis.com/v1/images:annotate?key=" . $ProductOpener::Config::google_cloud_vision_api_key;
+	# alpha-vision.googleapis.com/
 
-		if ($res->is_success) {
+	my $ua = LWP::UserAgent->new();
 
-			$log->info("request to google cloud vision was successful") if $log->is_info();
+	open(my $IMAGE, "<", $image_path) || die "Could not read $image_path: $!\n";
+	binmode($IMAGE);
+	local $/;
+	my $image_data = do {local $/; <$IMAGE>};    # https://www.perlmonks.org/?node_id=287647
+	close $IMAGE;
 
-			open(my $OUT, ">>:encoding(UTF-8)", "$data_root/logs/cloud_vision.log");
-			print $OUT "success\t" . $image_url . "\t" . $res->code . "\n";
-			close $OUT;
+	my $api_request_ref = {
+		requests => [
+			{
+				features => [
+					{type => 'TEXT_DETECTION'},
+					{type => 'LOGO_DETECTION'},
```

Comment:
for cropped images for ingredients and nutrition, we just need the text detection, maybe we could add an extra parameter to send_image_to_cloud_vision() so that we can state which detection we want.

Reply:
Oh, I see, I though it was not intended (as code was duplicated).

Reply:
We pay for each type of detection, so if we just need the text, then we should remove the others (which have been done on the full uncropped image already)